### PR TITLE
Provide access to TLS client certificate details

### PR DESCRIPTION
**Pull request by kentonv**
Decided to do this the "right way" by expanding the KJ networking interfaces with features for identifying the peer in general.

### CODE REVIEW

**Comment by kentonv** [comment]
Yay I broke all the builds. Looks like I used C++17 features and also forgot to update the Windows networking code...

File: c++/src/kj/async-io.h [hunk 1]
```diff
@@ -299,13 +299,115 @@ Own<AsyncIoStream> newPromisedStream(Promise<Own<AsyncIoStream>> promise);
...
+  // specific assumptions the caller can make about the type of `PeerIdentity`s used, allowing for
+  // identities to be statically downcast if the right conditions are met. In the absence of
+  // documented promises, RTTI may be needed to query the type.
+};
+
+class NetworkPeerIdentity: public PeerIdentity {
+  // PeerIdentity used for network protocols like TCP/IP. This identifies the remote peer.
+  //
+  // This is only "authenticated" to the extent that we know data written to the stream will be
+  // routed to the given address. This doesn not preclude the possibility of man-in-the-middle
```
Comment (jclee) [inline c++/src/kj/async-io.h:343]: "doesn not"

File: c++/src/kj/async-io.h [hunk 2]
```diff
@@ -299,13 +299,115 @@ Own<AsyncIoStream> newPromisedStream(Promise<Own<AsyncIoStream>> promise);
...
+  Own<AsyncIoStream> stream;
+  // The byte stream.
+
+  Own<PeerIdentity> peerIdentity;
+  // An object indicating who is at the other end of the stream.
+  //
+  // Different subclasses of `PeerIdentity` are used in different situations:
+  // - TCP connections will use NetworkPeerIdentity, which gives the network address of the client.
+  // - Local (unix) socket connections will use LocalPeerIdentity, which identifies the UID, GID,
+  //   and PID of the process that initiated the connection.
```
Comment (bretthoerner) [inline c++/src/kj/async-io.h:328]: nit: You say `GID` here but it looks like you nixed that after trying to implement it?

File: c++/src/kj/compat/tls.c++ [hunk 3]
```diff
@@ -845,6 +888,51 @@ TlsCertificate::~TlsCertificate() noexcept(false) {
...
+  KJ_ASSERT(index != -1, "certificate has no common name?");
+  X509_NAME_ENTRY* entry = X509_NAME_get_entry(subj, index);
+  KJ_ASSERT(entry != nullptr);
+  ASN1_STRING* data = X509_NAME_ENTRY_get_data(entry);
+  KJ_ASSERT(data != nullptr);
+
+  unsigned char* out = nullptr;
+  int len = ASN1_STRING_to_UTF8(&out, data);
+  KJ_ASSERT(len >= 0);
+  KJ_DEFER(OPENSSL_free(out));
```
Comment (bretthoerner) [inline c++/src/kj/compat/tls.c++:925]: huh, TIL OpenSSL has its own allocator. I guess that makes sense.

**Pull request merged**